### PR TITLE
feat(columnar): consumer value sub-columns via columnar ingest

### LINKED ISSUE #522: Consumer-provided columnar value layout: within-value sub-columns via caller ColumnBatch
## Problem

#503 lands the columnar PAX block format with the engine's **intrinsic-field columns** (`user_key`, `seqno`, `value_type`, `value`), where `value` is one opaque column. True analytical scans ("read a few of many columns without decoding whole rows") need the VALUE split into per-field sub-columns. A semantics-free engine cannot derive sub-columns from an opaque value without parsing it, which would put relational / graph semantics in the engine (a hard ban).

## Approach

The **consumer owns the column structure** (it differs per modality: time-series timestamp + metric columns, vector dimensions, document fields). The consumer transposes its rows into a `ColumnBatch` at its OWN flush / batch boundary (not the engine, and not on the write hot path) and hands it to the engine via a columnar ingest path. The engine stores the caller-supplied `ColumnBatch` as a #503 columnar block, applies per-column codecs + #502 zone-maps, and returns `ColumnBatch` on scan. The engine stays 100% semantics-free; column meaning lives in the consumer.

Precedent: schema-above-the-engine is standard for KV-on-LSM multimodal stores (YugabyteDB packed rows; "Columnar Formats for Schemaless LSM-based Document Stores", arxiv 2111.11517). WiscKey-style value opaqueness at the engine layer is preserved; this only adds an explicit columnar ingest path for callers that DO know their value structure.

## Contract

- A columnar ingest API: the consumer writes a `ColumnBatch` (its per-modality value sub-columns); the engine stores it as a #503 columnar block.
- A columnar scan returning `ColumnBatch` with projection (decode only referenced sub-columns) — this is the target of #506's full column projection once value sub-columns exist.
- The engine never parses values; it stores / retrieves typed opaque column chunks.

## Open decisions (resolve before code)

- Ingest API shape: a dedicated `ingest_columnar(ColumnBatch)` write path vs a value-class carrying a pre-transposed payload.
- How positional MVCC (#507 delete-bitmap) and seqno snapshot visibility apply over consumer-column rows.

## Dependency

#503 (columnar block format + `ColumnBatch`). Land AFTER #503's intrinsic-column foundation.

## no-std tier

Decode / scan read-path = core + alloc; ingest / transpose is the consumer's concern.

## Estimate

RFC ~3-4 days; implementation ~2-3 weeks.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Lets a consumer split a value into typed sub-columns and ingest a pre-transposed `ColumnBatch` directly, so analytical scans project individual sub-columns without decoding the whole value. The engine stays value-agnostic: it stores the sub-columns verbatim and reconstructs a row's value through a self-describing framing, never interpreting cell bytes.

## What it adds

- **Reversible value framing** (`frame_value_cells` / `unframe_value_cells`): a row's value sub-cells join into one self-describing blob (fixed-width cells verbatim, variable-width length-prefixed) that the consumer reverses by type tag. Fixed sub-columns carry no per-cell overhead.
- **Row reconstruction**: the columnar untranspose now accepts the three intrinsic columns plus one or more value sub-columns; a single value column round-trips byte-for-byte, two or more are framed. All existing corruption guards are preserved.
- **Writer + ingest path**: `Writer::write_columnar_batch` stores a consumer batch as one columnar block (keeping the sub-columns), wired through `MultiWriter` up to the public `tree.ingestion()?.write_columnar_batch(...).finish()`. The ingestion enables the columnar layout when the runtime config asks for it; a row-mode or blob tree rejects the batch.
- **Projection + MVCC composition**: the vectorized `columnar_scan(projection, predicate)` projects individual sub-columns, and the positional delete-bitmap masks sub-column rows in both point and projection reads (the mask is value-agnostic).

## v1 constraints

One batch is one block, keys must be sorted, value sub-columns are non-nullable, and per-row seqnos are `0` (the ingestion assigns the atomic global sequence number). Nullable value sub-columns and custom-comparator sorted-input validation are follow-ups.

## Testing

- Unit: framing round-trip plus edge cases, N-sub-column untranspose, decode guards.
- Integration (`tests/columnar_ingest.rs`): ingest round-trip through the public API, layout-required rejection, per-sub-column projection.
- Composition: the delete-bitmap masks value sub-columns in point and projection reads.
- Full gate: clippy `--all-features --all-targets`, no-std (`thumbv7em` + `alloc`) errcount 0, nextest (columnar 2163 + lz4,zstd 2104), doctests 61.

Closes #522

## Release Notes

* **New Features**
  * Added feature-gated columnar batch ingest for pre-transposed batches, including validation of intrinsic/value sub-column layout, strict key ordering, and proper behavior for empty batches.
  * Added utilities for framing/unframing multi-sub-column values to support correct reconstruction and projection/scans.
* **Bug Fixes**
  * Improved columnar batch encoding/recovery to correctly account for tombstones/weak-tombstones, deletes masking, and seqno bounds.
* **Tests**
  * Expanded end-to-end columnar ingest tests covering round-trips, projections, ordering/seqno rejection, and unsupported configurations.
* **Chores**
  * Reduced benchmark workflow timeout and shortened benchmark timing windows.

### CODE REVIEW

File: src/tree/ingest.rs [hunk 1]
```diff
@@ -277,6 +281,39 @@ impl<'a> Ingestion<'a> {
...
+        batch: &crate::table::columnar::ColumnBatch,
+    ) -> crate::Result<()> {
+        // Carry the batch's last key forward: it both records the ordering
+        // boundary for any later write and signals `finish` that data was
+        // written (it installs nothing when `last_key` is `None`).
+        if let Some(last) = self.writer.write_columnar_batch(batch)? {
+            self.last_key = Some(last);
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/tree/ingest.rs:330]: Fixed in 6ba0bc93b. Added an ingestion-level guard: a columnar batch whose first key does not strictly follow the last key written through the ingestion is rejected (Error::InvalidHeader), mirroring the row write methods. It persists across writer table rotations (where the inner writer running-key check resets), so an interleaved write then earlier-key batch errors instead of producing an unsorted run. Added ColumnBatch::first_user_key plus a regression test.

File: src/table/writer/mod.rs [hunk 2]
```diff
@@ -1108,10 +1135,150 @@ impl Writer {
...
+        }
+
+        self.encode_columnar_batch_block(
+            batch,
+            last_key.clone(),
+            last_seqno,
+            seqno_bounds,
+            item_count,
+            zone_block_min,
         )?;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/writer/mod.rs:1281]: @coderabbitai Fixed in 93d5d6ade. write_columnar_batch now calls self.spill_block() (after validation) before its metadata/locator accounting, so a buffered row chunk is registered first and the sorted block/index order + locator ordinals stay correct. Added a regression test (write a row, then a columnar batch, then finish; the row-written key now survives).

File: src/table/columnar.rs [hunk 3]
```diff
@@ -340,6 +340,28 @@ impl Column {
...
+            return Ok(None);
+        }
+        let key_col = self
+            .columns
+            .first()
+            .filter(|c| c.column_id == COL_USER_KEY)
+            .ok_or(Error::InvalidHeader(
+                "columnar: first column is not the user-key column",
+            ))?;
+        bytes_column_row(&key_col.data, self.row_count, 0).map(Some)
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/columnar.rs:371]: @coderabbitai Fixed in 85711d0a0. first_user_key now confirms the first column is the non-null Bytes user-key column and framing-checks it (key_col.validate) before the low-level bytes_column_row read, since it runs before the full untranspose validation. Added a test for a non-Bytes / missing key column.

File: src/table/columnar.rs [hunk 4]
```diff
@@ -717,28 +789,189 @@ pub fn column_batch_to_entries(batch: &ColumnBatch) -> Result<Vec<InternalValue>
...
+                let len = u32::try_from(cell.len()).map_err(|_| {
+                    Error::InvalidHeader("columnar: framed value sub-cell exceeds u32")
+                })?;
+                out.extend_from_slice(&len.to_le_bytes());
+                out.extend_from_slice(cell);
+            }
+        }
+    }
+    Ok(out)
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/columnar.rs:884]: @coderabbitai Fixed in 85711d0a0. frame_value_cells now rejects a fixed-width cell whose length does not equal its TypeTag width (it would misframe and shift later cells), returning InvalidHeader. Added a regression test.

File: tests/columnar_ingest.rs [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    )
+    .with_kv_separation(Some(Default::default()))
+    .open()?;
+
+    let mut ingest = any.ingestion()?;
+    assert!(
+        ingest
+            .write_columnar_batch(&one_fixed_subcolumn_batch(b"k0", 0))
+            .is_err(),
+        "columnar ingest is not supported on a blob tree",
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/columnar_ingest.rs:364]: @coderabbitai Fixed in 06b1e10a7. The blob-tree rejection test now matches Err(Error::FeatureUnsupported(_)) instead of .is_err(); also strengthened the other ingest-rejection tests to assert their exact variant (FeatureUnsupported for non-columnar layout / non-zero seqno, InvalidHeader for unsorted / out-of-order keys).

**Comment by polaz** [comment]
@coderabbitai review

**Comment by polaz** [comment]
@coderabbitai re-review

File: src/table/columnar.rs [hunk 6]
```diff
@@ -666,38 +697,88 @@ pub fn entries_to_column_batch(entries: &[InternalValue]) -> Result<ColumnBatch>
...
+    // Value sub-columns are consumer-defined (id / type / count). Nullable value
+    // sub-columns are a later slice, so reject a validity bitmap for now; the
+    // per-column `validate` bounds each against `row_count` like the intrinsics.
+    for col in value_cols {
+        if col.validity.is_some() {
+            return Err(Error::InvalidHeader(
+                "columnar: nullable value sub-columns are not supported yet",
+            ));
+        }
+        col.validate(batch.row_count)?;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/columnar.rs:790]: @coderabbitai Fixed in c21d8f576 (regression test in a1cb58d30). The untranspose now rejects a batch where any value sub-column id is below COL_VALUE (overlaps an intrinsic) or duplicates another value sub-column id, since projection selects by id.

File: src/table/tests.rs [hunk 7]
```diff
@@ -3913,3 +3921,277 @@ fn delete_bitmap_masks_deleted_key_in_point_read() -> crate::Result<()> {
...
+    let (_, checksum) = writer.finish()?.expect("table written");
+    let table = recover_test_table(&file, checksum)?;
+
+    // is_tombstone covers Tombstone + WeakTombstone; weak count is just the latter.
+    assert_eq!(table.metadata.tombstone_count, 2, "two tombstone-kind rows");
+    assert_eq!(table.metadata.weak_tombstone_count, 1, "one weak tombstone");
+    assert!(
+        table.get(b"k0", SeqNo::MAX, hash64(b"k0"))?.is_some(),
+        "the live row reads back",
+    );
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/tests.rs:4154]: @coderabbitai Fixed in 8230d0ba0. The test now asserts table.metadata.seqnos == (0, 0), proving the seqno-bounds section round-trips rather than relying on the point read alone.

File: src/table/tests.rs [hunk 8]
```diff
@@ -3913,3 +3921,277 @@ fn delete_bitmap_masks_deleted_key_in_point_read() -> crate::Result<()> {
...
+        .use_columnar(true)
+        .use_zone_map(true);
+    assert!(
+        writer
+            .write_columnar_batch(&empty, &crate::comparator::default_comparator())?
+            .is_none(),
+        "an empty batch yields no last key",
+    );
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/tests.rs:4213]: @coderabbitai Fixed in 8230d0ba0. The empty-batch test now also asserts writer.finish()?.is_none(), proving no SST is produced (not just no last key).

File: tests/columnar_ingest.rs [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    );
+
+    // Project only the bytes sub-column (id 4): again decoded in isolation.
+    let batches = table.columnar_scan(&[4], None)?;
+    for b in &batches {
+        assert!(
+            b.columns.iter().all(|c| c.column_id == 4),
+            "a sub-column-4 projection must not decode any other column",
+        );
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/columnar_ingest.rs]: @coderabbitai Fixed in 8230d0ba0. The bytes-projection assertion now collects the projected column data and compares it to the expected offset array [0,2,5] + payload "aabbb", not just the column id.

**Comment by polaz** [comment]
@coderabbitai review

**Comment by polaz** [comment]
@coderabbitai review
